File: momentra/momentra/Shell/BusinessActive/BusinessTabDataCache.swift
import Foundation

/// In-memory SWR cache for Business tab datasets keyed by momentId.
enum BusinessTabDataCache {
    struct PulseTab {
        let pulse: APIClient.BusinessPulsePayload
        let finance: APIClient.BusinessFinancePayload?
        let life: APIClient.BusinessLifePayload?
        let activities: [APIClient.ActivityItemPayload]
        let businessFamily: String?
        let facetStatus: String?
        let capacity: APIClient.BusinessCapacityPayload?
        let workload: APIClient.BusinessWorkloadPayload?
    }

    struct MemoryTab {
        let memory: APIClient.BusinessMemoryPayload?
        let pulse: APIClient.BusinessPulsePayload?
        let finance: APIClient.BusinessFinancePayload?
        let life: APIClient.BusinessLifePayload?
    }

    private static var pulseByMoment: [String: PulseTab] = [:]
    private static var memoryByMoment: [String: MemoryTab] = [:]

    static func peekPulse(_ momentId: String) -> PulseTab? {
        pulseByMoment[momentId]
    }

    static func putPulse(_ momentId: String, _ tab: PulseTab) {
        pulseByMoment[momentId] = tab
    }

    static func peekMemory(_ momentId: String) -> MemoryTab? {
        memoryByMoment[momentId]
    }

    static func putMemory(_ momentId: String, _ tab: MemoryTab) {
        memoryByMoment[momentId] = tab
    }

    /// Drop cached facets for one moment after a write so SWR reloads fresh data.
    static func invalidateMoment(_ momentId: String) {
        pulseByMoment.removeValue(forKey: momentId)
        memoryByMoment.removeValue(forKey: momentId)
    }

    static func clear() {
        pulseByMoment.removeAll()
        memoryByMoment.removeAll()
    }
}

enum BusinessTabPrefetch {
    static let activityLimit = 5

    /// Warm bundled pulse (finance + activity preview) so Business tabs paint without spinners.
    static func run(momentId: String) async {
        guard !momentId.isEmpty else { return }
        do {
            _ = try await BusinessTabLoad.loadPulseTab(momentId: momentId)
        } catch {
            // Prefetch is best-effort; visible tabs retry on their own .task.
        }
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/business/shared/BusinessTabDataCache.kt
package com.example.momentra.ui.shell.business.shared

import com.example.momentra.data.api.ActivityItemDto
import com.example.momentra.data.api.BusinessFinancePayloadDto
import com.example.momentra.data.api.BusinessLifePayloadDto
import com.example.momentra.data.api.BusinessMemoryPayloadDto
import com.example.momentra.data.api.BusinessPulsePayloadDto
import com.example.momentra.data.api.CapacityDto
import com.example.momentra.data.api.WorkloadDto
import java.util.concurrent.ConcurrentHashMap

/** In-memory SWR cache for Business tab datasets keyed by momentId. */
object BusinessTabDataCache {
    data class PulseTab(
        val pulse: BusinessPulsePayloadDto?,
        val finance: BusinessFinancePayloadDto?,
        val life: BusinessLifePayloadDto?,
        val activities: List<ActivityItemDto>,
        val businessFamily: String?,
        val facetStatus: String?,
        val capacity: CapacityDto? = null,
        val workload: WorkloadDto? = null,
    )

    data class MemoryTab(
        val memory: BusinessMemoryPayloadDto?,
        val pulse: BusinessPulsePayloadDto?,
        val finance: BusinessFinancePayloadDto?,
        val life: BusinessLifePayloadDto?,
    )

    private val pulseByMoment = ConcurrentHashMap<String, PulseTab>()
    private val memoryByMoment = ConcurrentHashMap<String, MemoryTab>()

    fun peekPulse(momentId: String): PulseTab? = pulseByMoment[momentId]

    fun putPulse(momentId: String, data: PulseTab) {
        pulseByMoment[momentId] = data
    }

    fun peekMemory(momentId: String): MemoryTab? = memoryByMoment[momentId]

    fun putMemory(momentId: String, data: MemoryTab) {
        memoryByMoment[momentId] = data
    }

    /** Drop cached facets for one moment after a write so SWR reloads fresh data. */
    fun invalidateMoment(momentId: String) {
        pulseByMoment.remove(momentId)
        memoryByMoment.remove(momentId)
    }

    fun clear() {
        pulseByMoment.clear()
        memoryByMoment.clear()
    }
}
